src/Middleware/ForwardedForMiddleware.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+namespace Amp\Http\Server\Middleware;
+
+use Amp\Cache\LocalCache;
+use Amp\Http;
+use Amp\Http\Server\Middleware;
+use Amp\Http\Server\Request;
+use Amp\Http\Server\RequestHandler;
+use Amp\Http\Server\Response;
+use Amp\Socket\CidrMatcher;
+use Amp\Socket\InternetAddress;
+use Amp\Socket\SocketException;
+
+final class ForwardedForMiddleware implements Middleware
+{
+    /** @var list<CidrMatcher> */
+    private readonly array $trustedProxies;
+
+    /** @var LocalCache<bool> */
+    private readonly LocalCache $trustedIps;
+
+    /**
+     * @param array<non-empty-string> $trustedProxies Array of IPv4 or IPv6 addresses with an optional subnet mask.
+     *      e.g., '172.18.0.0/24'
+     */
+    public function __construct(array $trustedProxies, int $cacheSize = 1000)
+    {
+        $this->trustedProxies = \array_map(
+            static fn (string $ip) => new CidrMatcher($ip),
+            \array_values($trustedProxies),
+        );
+
+        $this->trustedIps = new LocalCache($cacheSize);
+    }
+
+    public function handleRequest(Request $request, RequestHandler $requestHandler): Response
+    {
+        $clientAddress = $request->getClient()->getRemoteAddress();
+
+        if ($clientAddress instanceof InternetAddress && $this->isTrusted($clientAddress)) {
+            $request->setAttribute(self::class, $this->getForwarded($request));
```

Comment:
If the request comes from a trusted proxy, we can usually trust its added value in the header, but not _all_ values in the header.

Reply:
How should we approach this? Only provide the first array element or wrap them into another object, perhaps providing a `getTrusted(): ForwardedFor` and `getUntrusted(): list<ForwardedFor>`?

Reply:
Probably only one element, yes. I think it'll have to be the last (or if that is another trusted proxy, the second last). Providing the others in a wrapping object sounds fine, too. They can be some indicator for humans in attack analysis, but should never be used for security decisions.

Reply:
Yes, last element which is not itself a trusted proxy. I think then it may be best to an object where only that IP is provided. The header value itself is always available in the request for logging, etc. if someone desires.

Reply:
Updated to return the last element which itself is not a trusted proxy.